### PR TITLE
Improve and Add Comprehensive Test Coverage [#71]

### PR DESCRIPTION
**Pull request by Paul-Taiwo**
This PR addresses the issue [#71]. Key changes include:

1. New Test Suites:

- Added client.test.ts to verify high-level registration/authentication flows via the WebAuthn API mocks.
- Updates server.test.ts covering signature verification, challenge checking, and data parsing on the server side.
- Update parsers.test.ts to ensure correct parsing of client data, authenticator data, and registration/authentication structures.

2. Refined Mocks & Setup:

- Centralized global mocks for navigator.credentials, PublicKeyCredential, crypto.subtle, etc. into a single Jest setup file, simplifying maintenance and reducing duplication.
- Used environment checks (Node vs. jsdom) to conditionally load mocks, ensuring tests run smoothly in both environments.

3. Enhanced Coverage:
- Added tests for edge cases (e.g., missing challenges, invalid base64url).
- Verified abort behavior for concurrent WebAuthn calls.
- Tested fallback logic for missing AAGUIDs and updated metadata.
- Test coverage up to  94.57%

image

### CODE REVIEW

**Comment by dagnelies** [comment]
That sounds awesome, can't wait to test it out 🚀 

I'll review / try it later because I'm short on time right now, but at first glance I like it very much. Thanks 🙌

**Comment by dagnelies** [comment]
Hi @Paul-Taiwo. I reviewed your PR right now, and I noticed you moved a significant chunk of code from the server to utils. Why? Could you please revert that change? This would break *lots* of code out there. It would be ideal if this PR focuses solely on the test suite, not affecting the code itself.

File: tsconfig.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
 
   },
   "include": [
-    "./src/**/*.ts"
+    "./src/**/*.ts",
+    "global.d.ts"
```
Comment (dagnelies) [inline tsconfig.json:19]: And why is "global.d.ts" added?

Comment (Paul-Taiwo) [reply tsconfig.json]: Good points! The `global.d.ts` file is purely for typing the global mock variables. Regarding the production build, the declaration file don’t actually emit any JavaScript or impact runtime code, so they won’t alter the production bundle. Still, I've moved it into the jest_config file to keep things tidy and avoid confusion.

**Comment by Paul-Taiwo** [comment]
> Hi @Paul-Taiwo. I reviewed your PR right now, and I noticed you moved a significant chunk of code from the server to utils. Why? Could you please revert that change? This would break _lots_ of code out there. It would be ideal if this PR focuses solely on the test suite, not affecting the code itself.

The reason for splitting the functions out of the server.ts file is jest can’t mock same-file function calls by default. That means if verifyAuthentication() and verifySignature() (for example) both live in server.ts, any test that tries to mock verifySignature() will still end up calling the real function instead of the mock, because the file references it directly in-memory.

**Comment by dagnelies** [comment]
Upon further inspection, there is really a lot that is mocked and real payloads not tested at all. In the end, it tests very little and the coverage is quite misleading.

**Comment by Paul-Taiwo** [comment]
okay, I will review the test again.

**Comment by dagnelies** [comment]
Well, better partial tests than nothing. It can still be improved later on. Thanks for the contribution
